### PR TITLE
Add missing doc and tests for multi-key commands

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**

### CODE REVIEW

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 1]
```diff
@@ -2136,6 +2174,20 @@ public void zrangestore_by_lex(BaseClient client) {
         assertTrue(executionException.getCause() instanceof RequestException);
     }
 
+    @SneakyThrows
+    @ParameterizedTest(autoCloseArguments = false)
+    @MethodSource("getClients")
+    public void zrangestore_same_slot(BaseClient client) {
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:2180]: is it ok to keep it here? zrangestore has 3 test cases - with different range types

**Comment by Yury-Fridlyand** [comment]
Should I add a note for `del`, `exists`, etc that they accept keys from any slots (and corresponding tests)?
https://github.com/redis-rs/redis-rs/blob/7e79e3a380a07eb0c1e559d9afa9152a87d2e50c/redis/src/cluster_routing.rs#L473-L474

File: java/client/src/main/java/glide/api/commands/ListBaseCommands.java [hunk 2]
```diff
@@ -281,6 +281,8 @@ CompletableFuture<Long> linsert(
      * checked in the order that they are given.<br>
      * Blocks the connection when there are no elements to pop from any of the given lists.
      *
+     * @apiNote When in cluster mode, all <code>keys</code> must map to the same <code>hash slot
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/ListBaseCommands.java:284]: there are 2 `@apiNode` tags - is it ok?

File: node/.prettierignore [hunk 3]
```diff
@@ -0,0 +1 @@
+rust-client
```
Comment (acarbonetto) [inline node/.prettierignore:1]: what files does it pretty?  The .json files?

File: java/client/src/main/java/glide/api/commands/GenericBaseCommands.java [hunk 4]
```diff
@@ -18,6 +18,7 @@ public interface GenericBaseCommands {
      * Removes the specified <code>keys</code> from the database. A key is ignored if it does not
      * exist.
      *
+     * @apiNote When in cluster mode, <code>keys</code> can map to any <code>hash slot</code>s.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/GenericBaseCommands.java:21]: Review comments say that we should say: 
> the command is being routed to multiple nodes if keys belong to different slots.

What about: 
```
     * @apiNote When in cluster mode, the command may route to multiple nodes when <code>keys</code> map to different <code>hash slot</code>s.
```

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 5]
```diff
@@ -119,6 +120,11 @@ public void unlink_multiple_keys(BaseClient client) {
 
         Long unlinkedKeysNum = client.unlink(new String[] {key1, key2, key3}).get();
         assertEquals(3L, unlinkedKeysNum);
+
+        // command can handle cross-slot case
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:124]: can't we just put all the cross-slot commands into a single unit test in java/integTest/src/test/java/glide/cluster/ClusterClientTests.java? 

Same for single-node commands?

File: node/.prettierignore (hunk 3 above)
Comment (acarbonetto) [reply node/.prettierignore:1]: This should be in a different PR

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 5 above)
Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:124]: will do once we merge all these commands to avoid extra mess

**Comment by Yury-Fridlyand** [comment]
Paused - waiting for other commands to complete

File: java/client/src/main/java/glide/api/commands/ListBaseCommands.java (hunk 2 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/ListBaseCommands.java:284]: not ok

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 1 above)
Comment (acarbonetto) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:2196]: cross slot and range type seem like separate arguments.  It doesn't seem necessary to write another set of 3 tests.

**Review by acarbonetto — APPROVED** [review]

**Comment by Yury-Fridlyand** [comment]
ready for review

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 6]
```diff
@@ -335,6 +335,7 @@ public void set_missing_value_and_returnOldValue_is_null(BaseClient client) {
     @ParameterizedTest(autoCloseArguments = false)
     @MethodSource("getClients")
     public void mset_mget_existing_non_existing_key(BaseClient client) {
+        // keys are from different slots
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:338]: If you add a prefix to the key you can guarantee different slots.

File: java/integTest/src/test/java/glide/cluster/CommandTests.java [hunk 7]
```diff
@@ -633,4 +641,56 @@ public void objectFreq() {
...
+                        "zinterstore",
+                        clusterClient.zinterstore("foo", new KeyArray(new String[] {"abc", "zxy", "lkn"}))),
+                Arguments.of("brpop", clusterClient.brpop(new String[] {"abc", "zxy", "lkn"}, .1)),
+                Arguments.of("blpop", clusterClient.blpop(new String[] {"abc", "zxy", "lkn"}, .1)),
+                Arguments.of("pfcount", clusterClient.pfcount(new String[] {"abc", "zxy", "lkn"})),
+                Arguments.of("pfmerge", clusterClient.pfmerge("abc", new String[] {"zxy", "lkn"})));
+    }
+
+    @SneakyThrows
+    @ParameterizedTest(name = "{0} cross slot")
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/cluster/CommandTests.java:672]:
```suggestion
    @ParameterizedTest(name = "{0} cross slot keys will throw RequestException")
```

File: java/integTest/src/test/java/glide/cluster/CommandTests.java [hunk 8]
```diff
@@ -633,4 +641,56 @@ public void objectFreq() {
...
+                Arguments.of("exists", clusterClient.exists(new String[] {"abc", "zxy", "lkn"})),
+                Arguments.of("unlink", clusterClient.unlink(new String[] {"abc", "zxy", "lkn"})),
+                Arguments.of("del", clusterClient.del(new String[] {"abc", "zxy", "lkn"})),
+                Arguments.of("mget", clusterClient.mget(new String[] {"abc", "zxy", "lkn"})),
+                Arguments.of("mset", clusterClient.mset(Map.of("abc", "1", "zxy", "2", "lkn", "3"))),
+                Arguments.of("touch", clusterClient.touch(new String[] {"abc", "zxy", "lkn"})));
+    }
+
+    @SneakyThrows
+    @ParameterizedTest(name = "{0} cross slot")
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/cluster/CommandTests.java:691]:
```suggestion
    @ParameterizedTest(name = "{0} cross slot keys are allowed")
```

File: python/python/glide/async_commands/core.py [hunk 9]
```diff
@@ -500,6 +500,9 @@ async def rename(self, key: str, new_key: str) -> TOK:
         meaning that in practice only keys that have the same hash tag can be reliably renamed in cluster.
         See https://redis.io/commands/rename/ for more details.
 
+        Note:
+            When in cluster mode, the command may route to multiple nodes when `keys` map to different `hash slots`.
```
Comment (acarbonetto) [inline python/python/glide/async_commands/core.py:504]: why is `hash slots` in backticks?  This is not related to Python. 
We should only backtick Python-related variables and terms.

File: python/python/tests/test_async_client.py [hunk 10]
```diff
@@ -2191,6 +2191,28 @@ async def test_pfadd(self, redis_client: TRedisClient):
             await redis_client.pfadd("foo", [])
 
 
+class TestMultiKeyCommandCrossSlot:
+    @pytest.mark.parametrize("cluster_mode", [True])
+    @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
+    async def test_multi_key_command_returns_cross_slot_error(
+        self, redis_client: RedisClusterClient
+    ):
+        # TODO brpop, bz*, zdiff, sdiff and others - all rest multi-key commands except ones tested below
```
Comment (acarbonetto) [inline python/python/tests/test_async_client.py:2200]: TODO is for this PR? 

- [ ] Complete TODO

File: python/python/tests/test_async_client.py [hunk 11]
```diff
@@ -2191,6 +2191,28 @@ async def test_pfadd(self, redis_client: TRedisClient):
...
+    @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
+    async def test_multi_key_command_routed_to_multiple_nodes(
+        self, redis_client: RedisClusterClient
+    ):
+        await redis_client.exists(["abc", "zxy", "lkn"])
+        await redis_client.unlink(["abc", "zxy", "lkn"])
+        await redis_client.delete(["abc", "zxy", "lkn"])
+        await redis_client.mget(["abc", "zxy", "lkn"])
+        await redis_client.mset({"abc": "1", "zxy": "2", "lkn": "3"})
+        # TODO touch
```
Comment (acarbonetto) [inline python/python/tests/test_async_client.py:2213]: TODO is for this PR? 
- [ ] complete TODO

Comment (Yury-Fridlyand) [reply python/python/tests/test_async_client.py:2213]: next, touch is not implemented yet

File: python/python/tests/test_async_client.py (hunk 10 above)
Comment (Yury-Fridlyand) [reply python/python/tests/test_async_client.py:2200]: these are not implemented yet

File: python/python/glide/async_commands/core.py (hunk 9 above)
Comment (Yury-Fridlyand) [reply python/python/glide/async_commands/core.py:504]: we do that in java... ok I'll remove
Do you want me to remove that in java too?

**Comment by Yury-Fridlyand** [comment]
CI fixed, ready for review (again)
